### ISSUE TITLE
Type casting of binary op arguments causes kernel slowdowns

### ISSUE THREAD

**Issue by matthiasdiener** [inducer/pytato#542]
Contrary to what I speculated in #538, the cause of our slowdowns in mirgecom is not the numpy type promotion itself, but the casting of operands for binary ops: https://github.com/inducer/pytato/blob/3fedafc22829365e73aa5afbdf2d0061a19a574a/pytato/utils.py#L235-L254

As an example, consider:

```python
import numpy as np
import pytato as pt
import pyopencl as cl

a = np.array([1, 2, 3], dtype=np.uint32)
b = np.int32(7)

res = pt.make_data_wrapper(a) - b

ctx = cl.create_some_context()
queue = cl.CommandQueue(ctx)

prg = pt.generate_loopy(res)

print(prg.program)

_, out = prg(queue)

print("pytato", out[0], out[0].dtype)
print("numpy", a - b, (a-b).dtype)
```

With current pytato `main`, this generates:

```
---------------------------------------------------------------------------
KERNEL: _pt_kernel
---------------------------------------------------------------------------
ARGUMENTS:
_pt_data: type: np:dtype('uint32'), shape: (3), dim_tags: (N0:stride:1), offset: <class 'loopy.typing.auto'> in aspace: global
_pt_out: type: np:dtype('int64'), shape: (3), dim_tags: (N0:stride:1) out aspace: global
---------------------------------------------------------------------------
DOMAINS:
{  :  }
{ [_pt_out_dim0] : 0 <= _pt_out_dim0 <= 2 }
---------------------------------------------------------------------------
INAME TAGS:
_pt_out_dim0: None
---------------------------------------------------------------------------
INSTRUCTIONS:
for _pt_out_dim0
  _pt_out[_pt_out_dim0] = cast(np:dtype('int64'), _pt_data[_pt_out_dim0]) + (-1)*cast(np:dtype('int64'), 7)  {id=_pt_out_store}
end _pt_out_dim0
---------------------------------------------------------------------------

pytato [-6 -5 -4] int64
numpy [-6 -5 -4] int64
```

With the argument casting disabled (line 250 above), this generates:

```
---------------------------------------------------------------------------
KERNEL: _pt_kernel
---------------------------------------------------------------------------
ARGUMENTS:
_pt_data: type: np:dtype('uint32'), shape: (3), dim_tags: (N0:stride:1), offset: <class 'loopy.typing.auto'> in aspace: global
_pt_out: type: np:dtype('int64'), shape: (3), dim_tags: (N0:stride:1) out aspace: global
---------------------------------------------------------------------------
DOMAINS:
{  :  }
{ [_pt_out_dim0] : 0 <= _pt_out_dim0 <= 2 }
---------------------------------------------------------------------------
INAME TAGS:
_pt_out_dim0: None
---------------------------------------------------------------------------
INSTRUCTIONS:
for _pt_out_dim0
  _pt_out[_pt_out_dim0] = _pt_data[_pt_out_dim0] + -7  {id=_pt_out_store}
end _pt_out_dim0
---------------------------------------------------------------------------

pytato [4294967290 4294967291 4294967292] int64
numpy [-6 -5 -4] int64
```

In the first case, there is an extra `(-1)*...` (this gets propagated to pyopencl as well), which I suspect is (a part of ?) the reason for these slowdowns.

Disabling the casting restores the previous performance for the kernels of our runs.

**Comment by kaushikcfd** [comment]
Is the transform pathway the same i.e. does the new casting lead to FusionArrayContext not kick in? (Since the expressions have changed and maybe our matching logic is brittle.)

**Comment by inducer** [comment]
- To add onto what (I think) @kaushikcfd is implying: I think it's highly unlikely that the mere presence of the casts will do much to influence execution speed, much less explain the massive slowdowns (10x IIRC?) being seen. Often the generated assembly is the same or quite similar: https://godbolt.org/z/dv4rPTzzh
- IIRC `index_lambda_to_hlo` is only used in axis tag propagation. (@kaushikcfd correct? or are there other places?) In any case, axis tag propagation should be rewritten to instead be based on generated `IndexLambda`s, I expect this will be far more robust, and simpler to boot. This would allow us to get rid of `index_lambda_to_hlo` entirely.

**Comment by kaushikcfd** [comment]
> I think it's highly unlikely that the mere presence of the casts will do much to influence execution speed

I agree.

> IIRC index_lambda_to_hlo is only used in axis tag propagation

Yep, and also for codegen in the JAX target.

>  I expect this will be far more robust, and simpler to boot

:warning:  It's quite complicated and a time sink (I remember spending ~1 week trying to figure out the raising problem with no luck.)

**Comment by matthiasdiener** [comment]
I compared the execution of our prediction driver with and without the argument type casting workaround I outlined above, and I could see no indicators of the fusion actx kicking in (i.e., no messages regarding fallbacks etc.). 

I did tinker a bit with `index_lambda_to_high_level_op`, and it appears to me it is doing the right thing: type casts get dropped (via the `TypeCastDropper`); it finds a subtraction op in the typecast-enabled case (dropping the `(-1)*` part), and a summation op in the typecast-disabled case. I'm not sure that makes a difference for the transformations

**Comment by matthiasdiener** [comment]
Here is a difference seen in `index_lambda_to_high_level_op` (this may be a red herring): many `inner_expr` operate on floats instead of ints with the type casting enabled. E.g.:

```diff
-inner_expr: _in0[_0, _1]*3
-inner_expr: _in0[_0, _1]*3
-inner_expr: _in0[_0, _1]*3
-inner_expr: _in0[_0, _1]*3
-inner_expr: _in0[_0, _1]*3
-inner_expr: _in0[_0, _1]*3
-inner_expr: _in0[_0, _1]*300
-inner_expr: _in0[_0, _1]*300
-inner_expr: _in0[_0, _1]*300
+inner_expr: _in0[_0, _1]*3.0
+inner_expr: _in0[_0, _1]*3.0
+inner_expr: _in0[_0, _1]*3.0
+inner_expr: _in0[_0, _1]*3.0
+inner_expr: _in0[_0, _1]*3.0
+inner_expr: _in0[_0, _1]*3.0
+inner_expr: _in0[_0, _1]*300.0
+inner_expr: _in0[_0, _1]*300.0
+inner_expr: _in0[_0, _1]*300.0
```

**Comment by matthiasdiener** [comment]
Looking through all the generated loopy kernels (at the end of `pt.generate_loopy`) shows that all kernels have identical code except for the types of scalar constants (as indicated by my previous comment). This seems to indicate that code transformations are not the reason for the slowdown?

The rhs kernels as well as a diff between them can be found [here](https://gist.github.com/matthiasdiener/273cacee6321093ed27dc6e2569528f4).

**Comment by kaushikcfd** [comment]
Looking at the kernels, doesn't look like they got transformed with loop_fusion. (It maybe possible that these weren't printed before generate_code_v2?)

**Comment by matthiasdiener** [comment]
Unfortunately, the differences are more serious after calling `lp.generate_code_v2`. I've made a new gist: https://gist.github.com/matthiasdiener/ab9f72df555711b2419903f5c9e8c5a9

**Comment by kaushikcfd** [comment]
Thanks! Except the power function (float**float vs float ** int), I don't see anything out of the ordinary.

**Comment by kaushikcfd** [comment]
For the earlier version of the `pow`s, the native compiler could CSE-out most of the terms.

**Comment by matthiasdiener** [comment]
>  An easy way to check this is just disabling the casting for pow terms and seeing if we can get back our earlier perf.

Yes, disabling casting only for pow (via `if op == prim.Power:`) restores the fast performance.

**Comment by kaushikcfd** [comment]
Thanks! This is a bug in potato. The proposed patch is fine by me.

Side note: Mirge-Com should probably use Horner's scheme for polynomial evaluation.

**Comment by matthiasdiener** [comment]
> Thanks! 

Thank you for your help! I'm thinking of restricting the workaround in #543 to only skip casting for `pow` with `int`-like arguments.
